Design APIs of the HeatMap widget

fgksgf opened the pull request:
Most of the design refers to the implementation of LineChart, but some simplifications have been made, such as:
- Remove some options
- Zooming is not supported (I don't think it is necessary)
- ...

Feel free to ask me if you have any questions :)

fgksgf commented:
@googlebot I signed it!

mum4k reviewed (changes requested):
This looks great! Thank you for sending a PR with a small enough scope to enable a productive discussion during the code review.

I am also impressed that the PR is already based off the `devel` branch and contains skeletons for all the internal libraries and the demo. Thank you for doing the additional work and research.

mum4k commented:
@fgksgf in addition to addressing the comments above, can you please also explicitly state that you give me permission to merge this PR and the remaining PRs implementing the HeatMap into the `master` branch?

This is required since this has to pass the CLA check one more time when I will be doing the release.

fgksgf commented on widgets/heatmap/heatmapdemo/heatmapdemo.go:
```diff
@@ -0,0 +1,64 @@
+)
+
+func main() {
+	t, err := termbox.New()
```
No problem.

fgksgf commented on widgets/heatmap/heatmap.go:
```diff
@@ -0,0 +1,124 @@
+// HeatMap draws heat map charts.
+//
+// Heatmap consists of several cells. Each cell represents a value.
+// The larger the value, the darker the color of the cell (from white to black).
```
Actually, I have tried red color, but I found that the effect was not ideal. The color difference between different values was not obvious. This is because the number of red in Xterm colors is limited (only about a dozen from light red to dark red). However there are more colors from white to gray to black colors (more than 20 kinds), which can better reflect the difference between the values.

So if you want to make the colors configurable, it requires the terminal supports 'true color'.

mum4k replied:
That is a good point, I understand now why you chose to use the shades of grey and it does make sense.

fgksgf commented on widgets/heatmap/internal/axes/axes.go:
```diff
@@ -0,0 +1,85 @@
+
+// NewYDetails retrieves details about the Y axis required
+// to draw it on a canvas of the provided area.
+func NewYDetails(stringLabels []string) (*YDetails, error) {
```
ok

fgksgf commented on widgets/heatmap/options.go:
```diff
@@ -0,0 +1,74 @@
+// options stores the provided options.
+type options struct {
+	// The default value is 3
+	cellWidth      int
```
Good idea

fgksgf commented on widgets/heatmap/heatmap.go:
```diff
@@ -0,0 +1,124 @@
+// Each value in values has a xLabel and a yLabel, which means
+// len(xLabels) == len(values) and len(yLabels) == len(values[i]).
+// Provided options override values set when New() was called.
+func (hp *HeatMap) Values(xLabels []string, yLabels []string, values [][]float64, opts ...Option) error {
```
What you said makes sense, though my design is similar to matlab and python's [1,2].

How about this?  Split the API into three:
``` go
func (hp *HeatMap) Values(values [][]float64, opts ...Option) error { ... }
func (hp *HeatMap) XLabels(xLabels []string, opts ...Option) error { ... }
func (hp *HeatMap) YLabels(yLabels []string, opts ...Option) error { ... }
```
Users can only set data, they don't have to specify labels, then labels will be generated automatically, like `{"0","1","2",...}`.

[1] https://www.mathworks.com/help/matlab/ref/heatmap.html#description
[2] https://seaborn.pydata.org/generated/seaborn.heatmap.html

mum4k replied:
Thanks for providing the API examples. I think your latest proposal is a good one and aligns with the two examples. I think we should also give users a method (or methods) to clear the labels. So they can both add them and remove them.

Are we still going to require that the number of labels matches the number of values on each axis?

One remaining question we should still answer is how should this behave in regards to the limited space available on screen / in the container. Let's say that the user provides values that have 100 data points on the X axis, but the container only provides the HeatMap widget with 50 cells on that axis. What should the widget do?

fgksgf replied:
>  I think we should also give users a method (or methods) to clear the labels. So they can both add them and remove them.

It makes sense.

> Are we still going to require that the number of labels matches the number of values on each axis?

I think so. For example, if there are six columns of data, uses have to provide six X labels, but some of them can be empty string.

> One remaining question we should still answer is how should this behave in regards to the limited space available on screen / in the container. Let's say that the user provides values that have 100 data points on the X axis, but the container only provides the HeatMap widget with 50 cells on that axis. What should the widget do?

Sorry, I'm not sure what you mean. Do you mean what should the widget do when the canvas is not big enough to draw all cells of the heat map ? If so, the heat map will not be drawn.

mum4k replied:
Thank you for explaining the behavior when the canvas is not big enough. The proposal sounds good, but we should consider giving the user a hint of how many values they can display. The users of the widget will likely prefer to aggregate the data or push less values, rather than pushing values that will result in nothing displayed. Can we include a function that will hint the user about the available space? There are a couple of examples in the other widgets, e.g. the BarChart's `ValueCapacity()` method:

https://github.com/mum4k/termdash/blob/fbd21e790423e0d479d440e6d1f77e1ca7f01382/widgets/barchart/barchart.go#L241-L256

Lastly after thinking about this widget more - it is also important to give the user a way of atomically updating both the values and the labels. This is because in the default execution mode, `termdash` controls when the screen is redrawn, the user doesn't control it. Now imagine the user wants to display a new set of values with a new set of labels. If we go with separate methods for updating labels and values, the screen could redraw in between causing the terminal to intermittently display inconsistent data.

I think your original proposal of having just one method that accepts the labels and the values was the best one. Sorry that I didn't think about this before. What do you think, should we go back to that proposal?

fgksgf replied:
> Can we include a function that will hint the user about the available space?

No problem.

>  What do you think, should we go back to that proposal?

You are right, I didn't think about updating the data either. Let's go back. 😂
Nevertheless, our discussion is still very meaningful and enjoyable, I have benefited a lot. Thank you.

fgksgf commented:
> @fgksgf in addition to addressing the comments above, can you please also explicitly state that you give me permission to merge this PR and the remaining PRs implementing the HeatMap into the `master` branch?
> 
> This is required since this has to pass the CLA check one more time when I will be doing the release.

I give @mum4k permission to merge this PR and the remaining PRs implementing the HeatMap into the master branch.

Like this ?

mum4k commented:
Thank you for stating the permission to merge, that statement is perfect @fgksgf.

mum4k commented on widgets/heatmap/options.go:
```diff
@@ -59,6 +59,13 @@ func (o option) set(opts *options) {
 	o(opts)
 }
 
+// CellWidth set the width of each cell.
```
It may be worth to spend a bit more time here, feels like this is an important detail about how HeatMap works / how it assigns data to cells, but the current description of this option doesn't really help me to understand it. A width of a terminal cell is fixed to just 1. What does it mean when we change it to a higher number? What impact does it have on how we are going to present the data?

fgksgf replied:
The cellWidth is the width of cells(or grids) in the heat map, not the terminal cell.
The default height of each cell (grid) is 1 and the width is 3.
In the figure below, the cell is framed in red：

<img width="799" alt="Jietu20201122-230412" src="https://user-images.githubusercontent.com/26627380/99907322-2731b980-2d17-11eb-9c3c-02ae35821803.png">

mum4k replied:
Thank you, I understand this now. Would you be able to update the comment on the option, so that it explains this for our users?

fgksgf replied:
OK

mum4k commented on widgets/heatmap/heatmap.go:
```diff
@@ -0,0 +1,153 @@
+}
+
+// ClearXLabels clear the X labels.
+func (hp *HeatMap) ClearXLabels() {
```
Similarly to why we wanted a single function to update the the labels and values (atomicity) we will likely need a single one to clear them. At this point I am wondering if we even need a dedicated clear function. The caller can just call `Values()` again with xLabels == yLabels == nil.

What do you think?

fgksgf replied:
>  Also looks like the build is failing due to a golint error, could you fix that before we move forward?

Done

Do you mean we need add a function, like this:
```golang
func (hp *HeatMap) ClearLabels() {
    hp.xLabels = nil
    hp.yLabels = nil
}
```

I think it's ok, better than nothing, because it will not affect the implementation of HeatMap.

mum4k replied:
Yes, we can either add a function like that, or we can just use `Values()`. Another call to values with both the `xLabels` and the `yLabels` slices empty should achieve the same result (clear labels, replace values).

Up to you really.

fgksgf replied:
I think just use `Values()` is ok.

mum4k reviewed (changes requested):
All looks good, just one comment about the Clear* methods. Thank you for the good discussion on this PR.

Also looks like the build is failing due to a golint error, could you fix that before we move forward? The error seems to be:

```
$ diff -u <(echo -n) <(golint ./...)
--- /dev/fd/63	2020-11-23 03:03:35.141897999 +0000
+++ /dev/fd/62	2020-11-23 03:03:35.141897999 +0000
@@ -0,0 +1 @@
+widgets/heatmap/internal/axes/axes.go:25:7: exported const AxisWidth should have comment or be unexported
The command "diff -u <(echo -n) <(golint ./...)" exited with 1.
```
